subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/InputPropertyAnnotationHandler.java

```diff
@@ -66,31 +70,57 @@ public void visitPropertyValue(String propertyName, PropertyValue value, Propert
+                        .withDescription(() -> "of type " + valueType.getName() + " shouldn't be annotated with @Optional")
+                        .happensBecause("Properties of primitive type cannot be optional")
+                        .addPossibleSolution("Remove the @Optional annotation")
+                        .addPossibleSolution(() -> "Use the " + JavaReflectionUtil.getWrapperTypeForPrimitiveType(valueType).getName() + " type instead")
+                        .documentedAt("validation_problems", "cannot_use_optional_on_primitive_types")
+            );
+        }
+    }
+
+    private void validateNotFileType(PropertyMetadata propertyMetadata, TypeValidationContext validationContext, Class<?> valueType) {
         if (File.class.isAssignableFrom(valueType)
+            || RegularFile.class.isAssignableFrom(valueType)
+            || RegularFileProperty.class.isAssignableFrom(valueType)
             || java.nio.file.Path.class.isAssignableFrom(valueType)
             || FileCollection.class.isAssignableFrom(valueType)) {
             validationContext.visitPropertyProblem(problem ->
                 problem.withId(ValidationProblemId.INCORRECT_USE_OF_INPUT_ANNOTATION)
-                    .forProperty(propertyMetadata.getPropertyName())
-                    .reportAs(ERROR)
-                    .withDescription(() -> String.format("has @Input annotation used on property of type '%s'", ModelType.of(valueType).getDisplayName()))
-                    .happensBecause(() -> "A property of type '" + ModelType.of(valueType).getDisplayName() + "' annotated with @Input cannot determine how to interpret the file")
-                    .addPossibleSolution("Annotate with @InputFile for regular files")
-                    .addPossibleSolution("Annotate with @InputDirectory for directories")
-                    .addPossibleSolution("If you want to track the path, return File.absolutePath as a String and keep @Input")
-                    .documentedAt("validation_problems", "incorrect_use_of_input_annotation")
+                        .forProperty(propertyMetadata.getPropertyName())
+                        .reportAs(ERROR)
+                        .withDescription(() -> String.format("has @Input annotation used on property of type '%s'", ModelType.of(valueType).getDisplayName()))
+                        .happensBecause(() -> "A property of type '" + ModelType.of(valueType).getDisplayName() + "' annotated with @Input cannot determine how to interpret the file")
+                        .addPossibleSolution("Annotate with @InputFile for regular files")
+                        .addPossibleSolution("Annotate with @InputFiles for collections of files")
+                        .addPossibleSolution("If you want to track the path, return File.absolutePath as a String and keep @Input")
+                        .documentedAt("validation_problems", "incorrect_use_of_input_annotation")
```

Comment:
Can you please revert the formatting changes here? The code should be indented as it was originally.

Reply:
Done.